tools/find_deprecated.py

```diff
@@ -0,0 +1,175 @@
+                    self._since = ".".join(kwarg.value.value.split(".")[:2])
+        return self._since
+
+    @property
+    def lineno(self):
+        """Line number of the decorator."""
+        return self.decorator_node.lineno
+
+    @property
+    def target(self):
+        """Name of the decorated function/method."""
+        return self.func_node.name
+
+    @property
+    def location_str(self):
+        """String with the location of the deprecated decorator <filename>:<line number>"""
+        return f"{self.filename}:{self.lineno}"
+
+
+class DecoratorVisitor(ast.NodeVisitor):
+    """
+    Node visitor for finding deprecation decorator
+    Args:
+        filename: Name of the file to analyze
+    """
+
+    def __init__(self, filename=None):
+        self.filename = filename
+        self.deprecations = []
+
+    @staticmethod
+    def is_deprecation_decorator(node):
+        """Check if a node is a deprecation decorator"""
+        return (
+            isinstance(node, ast.Call)
+            and isinstance(node.func, ast.Name)
+            and node.func.id.startswith("deprecate_")
+        )
+
+    def visit_FunctionDef(self, node):  # pylint: disable=invalid-name
+        """Visitor for function declarations"""
+        self.deprecations += [
+            Deprecation(self.filename, d_node, node)
+            for d_node in node.decorator_list
+            if DecoratorVisitor.is_deprecation_decorator(d_node)
+        ]
+        ast.NodeVisitor.generic_visit(self, node)
+
+
+class DeprecationCollection:
+    """
+    A collection of :class:~.Deprecation
+
+    Args:
+        dirname: Directory name that would be checked recursively for deprecations.
+    """
+
+    def __init__(self, dirname):
+        self.dirname = dirname
+        self._deprecations = None
+        self.grouped = OrderedDict()
+
+    @property
+    def deprecations(self):
+        """List of deprecation :class:~.Deprecation"""
+        if not self._deprecations:
+            self.collect_deprecations()
+        return self._deprecations
+
+    def collect_deprecations(self):
+        """Run the :class:~.DecoratorVisitor on `self.dirname`"""
+        self._deprecations = []
+        files = [self.dirname] if self.dirname.is_file() else self.dirname.rglob("*.py")
+        for filename in files:
+            self._deprecations.extend(DeprecationCollection.find_deprecations(filename))
+
+    def group_by(self, attribute_idx):
+        """Group :class:~`.Deprecation` in self.deprecations based on the attribute attribute_idx"""
+        grouped = defaultdict(list)
+        for obj in self.deprecations:
+            grouped[getattr(obj, attribute_idx)].append(obj)
+        for key in sorted(grouped.keys()):
+            self.grouped[key] = grouped[key]
+
+    @staticmethod
+    def find_deprecations(file_name):
+        """Runs the deprecation finder on file_name"""
+        with open(file_name, encoding="utf-8") as fp:
+            code = fp.read()
```

Comment:
FYI `Path(file_name).read_text()` is more concise.

Reply:
good point. Done in 1ecc5afe64a93f476dfeb922de42f7160f2bc8ce (with you as co-author)